### PR TITLE
FIP-0088: Add support for upgradable actors

### PR DESCRIPTION
**Pull request by fridrik01**
Discussion: https://github.com/filecoin-project/FIPs/discussions/396

See https://github.com/filecoin-project/ref-fvm/issues/717
Implementation in https://github.com/filecoin-project/ref-fvm/pull/1866

### CODE REVIEW

File: FIPS/fip-actor-upgrades.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+## Abstract
+
+This FIP proposes the integration of upgradable actors into the Filecoin network through the introduction of a new `upgrade_actor` syscall and an optional `upgrade` WebAssembly (WASM) entrypoint.
+
+Upgradable actors provide a framework for seamlessly replacing deployed actor code, significantly enhancing the user experience when updating deployed actor code.
+
+## Change Motivation
+
+Currently, all actors on the Filecoin Network are immutable once deployed. To modify the underlying actor code, such as fixing a security bug, the following steps are required:
```
Comment (Stebalien) [inline FIPS/fip-actor-upgrades.md:23]: Nits: the code associated with an actor is immutable, the actor itself is mutable.

File: FIPS/fip-actor-upgrades.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+1. Deploy a new actor with the corrected code.
+2. Migrate all state from the previous actor to the new one.
+3. Update all other actors interacting with the old actor to use the new actor.
+
+By adding support for upgradable actors, deployed actors can easily upgrade their code and no longer need to go through the series of steps mentioned above.
+
```
Comment (Stebalien) [inline FIPS/fip-actor-upgrades.md:29]: I'd add that this FIP is _immediately_ motivated by https://github.com/filecoin-project/FIPs/discussions/473#discussioncomment-4774001.

File: FIPS/fip-actor-upgrades.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+This also simplifies the upgrade process for builtin actors. Currently, any changes made to the builtin actors require a network upgrade, a new Lotus version, and collaboration with the network to push the new version to production. This is a time consuming and manual process. However, with upgradable actors, we can instead deploy a new code to the builtin actors allowing the changes to be pushed to production almost instantly and does not require a new network upgrade or new Lotus version.
```
Comment (Stebalien) [inline FIPS/fip-actor-upgrades.md:30]: This isn't true, unfortunately. We still need to do this out of band because:

1. It lives outside of consensus. Nobody has the authority to "change" these actors, so the entire network needs to simultaneously decide to replace the old code with new code.
2. It needs to happen in parallel with lots of caching (would take _way_ too long otherwise).

I'd instead say that this paves the way for moving _some_ network upgrade logic on-chain in the future.

File: FIPS/fip-actor-upgrades.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+### New upgrade_actor syscall
+
+We introduce a new `upgrade_actor` syscall which calls the `upgrade` wasm entrypoint of the calling actor and then atomically replaces the code CID of the calling actor with the provided code CID. It is defined as follows:
+
+```rust
+pub fn upgrade_actor(
+    new_code_cid_off: *const u8,
+    params: u32,
+) -> Result<Send>;
```
Comment (Stebalien) [inline FIPS/fip-actor-upgrades.md:75]: I'd define `Send` here, and rename it to `CallResponse`.

File: FIPS/fip-actor-upgrades.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+- `new_code_cid_off`: The code CID the calling actor should be replaced with.
+- `params`: The IPLD block handle passed, or `0` for none.
+
+On successful upgrade, this syscall will not return. Instead, the current invocation will "complete" and the return value will be the block returned by the new code's `upgrade` endpoint. If the new code rejects the upgrade (calls `sdk::vm::exit()`) or performs an illegal operation, this syscall will return the exit code plus the error returned by the upgrade endpoint.
+
+This syscall will:
+1. Validate that the pointers passed to the syscall are in-bounds.
+2. Validate that `new_code_cid_off` is a valid code CID.
+3. Validate that the calling actor is not currently executing in "read-only" mode. If so, the syscall fails with a "ReadOnly" (13) syscall error.
+4. Checks whether the calling actor is already on the call stack where it has previously been called on its `invoke` entrypoint (note that we allow calling `upgrade` recursively). If so, the syscall fails with a "Forbidden" (11) syscall error.
```
Comment (Stebalien) [inline FIPS/fip-actor-upgrades.md:88]: I'd give an example here.

**Comment by jennijuju** [comment]
Peer q: how can we leverage this with buitin actors? anyways to simplify network upgrade process when simple actor code change involved?

File: FIPS/fip-actor-upgrades.md [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+Detailed tests cover these security considerations and edge cases.
+
+## Incentive Considerations
+
+This FIP does not materially impact incentives in any way.
+
+## Product Considerations
+
+This FIP makes it possible to upgrade deployed actors, for example in cases where a bug or security concern was identified in the deployed code, allowing a simple safe way to address such issues which significantly improves the user experience from how it is today.
```
Comment (jennijuju) [inline FIPS/fip-actor-upgrades.md:141]: editor: probably worth discussing the immutability characteristic of the user actors (smart contracts) and why the ability of changing user actors are desirable on network like Filecoin, while its still less common in other smart contract platforms.

Comment (jennijuju) [reply FIPS/fip-actor-upgrades.md:141]: okay to move this discussion to the FIP discussion

File: FIPS/fip-actor-upgrades.md [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+2. Adding a new `upgrade_actor` syscall, enabling actors to upgrade themselves.
+
+These changes are discussed in detail in the following sections.
+
+### New upgrade WASM Entrypoint
+
+We introduce a new optional `upgrade` WASM entrypoint. Deployed actors must implement this entrypoint to be a valid upgrade target. It is defined as follows:
+
+```rust
+pub fn upgrade(params_id: u32, upgrade_info_id: u32) -> u32
```
Comment (jennijuju) [inline FIPS/fip-actor-upgrades.md:48]: I think an FRC for upgradable actors is worth being proposed along with this FIP, so that other FRCs can be built on top of it (with extended functionalities). similarly to https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/master/contracts/

File: FIPS/fip-actor-upgrades.md [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+5. Checks that we have space for storing the return block. If not, the syscall fails with a "LimitExceeded" (3) syscall error.
+6. Start a new Call Manager transaction:
+    1. Validate that the calling actor has not been deleted. If so, the syscall fails with a "IllegalOperation" (2) syscall error.
+    2. Update the actor in the state tree with the new `new_code_cid` keeping the same `state`, `sequence` and `balance`.
+    3. Invoke the target actor's `upgrade` entrypoint.
+    4. If the target actor does not implement the `upgrade` entrypoint, the syscall fails with a `ExitCode::SYS_INVALID_RECEIVER` exit code.
+    5. If the target actor aborts the `upgrade` entrypoint by calling `sdk::vm::exit()`, the syscall fails with the provided exit code.
+7. Apply transaction, committing changes.
```
Comment (jennijuju) [inline FIPS/fip-actor-upgrades.md:108]: does the calling actor pays for the transaction?

File: FIPS/fip-actor-upgrades.md [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+Parameters:
+- `new_code_cid_off`: The code CID the calling actor should be replaced with.
+- `params`: The IPLD block handle passed, or `0` for none.
+
+The `CallResult` struct is defined as follows:
+
+```rust
+pub struct CallResult {
+    pub block_id: BlockId,
```
Comment (jennijuju) [inline FIPS/fip-actor-upgrades.md:88]: sorry what's block id and block stat here ?

File: FIPS/fip-actor-upgrades.md [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+
+## Design Rationale
+
+### Additional metadata syscall
+
+We considered adding a new `get_old_code_cid` syscall to get the calling actors code CID. That has the benefit of keeping the `upgrade` entrypoint signature consistent with the `invoke` signature. We however rejected that as we felt the benefit didn't outweigh the overhead of adding a new syscall. Furthermore, it did not provide the flexibility of passing in an IPLD handle for a `UpgradeInfo` struct where we can easily add more fields if required.
+
+## Backwards Compatibility
+
+Fully backwards compatibility is expected.
```
Comment (jennijuju) [inline FIPS/fip-actor-upgrades.md:119]:
```suggestion
Full backwards compatibility is expected.
```

**Comment by jennijuju** [comment]
editor: FIP itself has all the components needed clearly written. Approving this as a DRAFT with one editorial clarification request. 

However, I think we need more discussion on whether this is a desired feature on Filecoin network & have better understanding on security implications for network application users when user actors aren't immutable. i.e: Defi and so on. See [here](https://github.com/filecoin-project/FIPs/pull/873#discussion_r1441702560)

**Review by jennijuju — APPROVED** [review]

File: FIPS/fip-actor-upgrades.md [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+## Change Motivation
+
+Currently, the code associated with all actors on the Filecoin Network are immutable once deployed. To modify the actor code, such as fixing a security bug, the following steps are required:
+1. Deploy a new actor with the corrected code.
+2. Migrate all state from the previous actor to the new one.
+3. Update all other actors interacting with the old actor to use the new actor.
+
+By adding support for upgradable actors, deployed actors can easily upgrade their code and no longer need to go through the series of steps mentioned above.
+
+This FIP is also motivated by the `f4` extensible address class which was introduced in [FIP-0048] and required special "placeholder" actors to support interactions with addresses that do not yet exist on-chain. With upgradable actors we can simplify this address class and remove these placeholder actors completely and instead deploy real actors and upgrade their code on first send.
```
Comment (kaitlin-beegle) [inline FIPS/fip-actor-upgrades.md:30]:
```suggestion
This FIP is also motivated by the `f4` extensible address class which was introduced in [FIP-0048] and required special "placeholder" actors to support interactions with addresses that do not yet exist on-chain. With upgradable actors we can simplify this address class and remove these placeholder actors completely. We will be able to deploy real actors and upgrade their code on first send.
```

File: FIPS/fip-actor-upgrades.md [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+
+Furthermore, this FIP paves the way for moving more network upgrade logic on-chain in the future, enabling a more seamless process for implementing critical updates and ensuring the continuous improvement of the Filecoin Network.
```
Comment (kaitlin-beegle) [inline FIPS/fip-actor-upgrades.md:32]: Perhaps a question for @jennijuju @steven004 and/or @lemmih, but I'm curious if there are implementation features that _should be_ coded in network actors but previously weren't because of the difficulty of migrating state for simple changes. 

Something to think about when/if we introduce upgradeable actors.

File: FIPS/fip-actor-upgrades.md [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+
+- Upgradable actors are opt-in by default, ensuring no impact on currently deployed actors.
+- Actors can only upgrade themselves, preventing one actor from upgrading another actor to a new version.
+- We reject re-entrant `upgrade_actor` syscalls, i.e., if some actor `A` is already on the call stack, no "deeper" instance of `A` should be able to call the upgrade syscall.
+
+Detailed tests cover these security considerations and edge cases.
+
+## Incentive Considerations
+
+This FIP does not materially impact incentives in any way.
```
Comment (kaitlin-beegle) [inline FIPS/fip-actor-upgrades.md:137]: Is this true? I'm curious how this change may impact the logic around timing and resource use for network upgrades, as well as the prioritization of certain changes. 

If actors' code can be much more efficiently upgraded, shouldn't we technically be able to push more updates/small changes more quickly?

**Review by kaitlin-beegle — APPROVED** [review]
[Editorial Review] - A few small nits and open questions, nothing major.  Editorially, this looks good to proceed.

File: FIPS/fip-actor-upgrades.md [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+author: Fridrik Asmundsson (@fridrik01), Steven Allen (@stebalien)
+discussions-to: https://github.com/filecoin-project/FIPs/discussions/396
+status: Draft
+type: Technical Core
+created: 2023-11-27
+---
+
+## Simple Summary
+
+This FIP introduces support for upgradable actors, enabling deployed actors to update their code while retaining their address, state, and balance.
```
Comment (Stebalien) [inline FIPS/fip-actor-upgrades.md:13]: Let's add a sentence explaining that:

1. This feature can currently only be used by built-in actors.
2. No actors currently use this feature.

Comment (maciejwitowski) [inline FIPS/fip-actor-upgrades.md:13]: @fridrik01 I think we should highlight that this FIP doesn't actually let anyone upgrade actors, at least the FEVM ones. Otherwise it may be confusing for the community members who will read it as "oh cool, upgrading actors is natively supported on Filecoin _now_".

Comment (Stebalien) [reply FIPS/fip-actor-upgrades.md:13]: Maybe also a sentence about how this is _not_ an on-chain upgrades feature?

Comment (maciejwitowski) [reply FIPS/fip-actor-upgrades.md:13]: This is the same comment as Steb's above https://github.com/filecoin-project/FIPs/pull/873#discussion_r1443105874

Comment (jennijuju) [reply FIPS/fip-actor-upgrades.md:13]: OH. Then my main questions can be resolved ….

File: FIPS/fip-actor-upgrades.md (hunk 7 above)
Comment (jennijuju) [reply FIPS/fip-actor-upgrades.md:48]: disregard - given https://github.com/filecoin-project/FIPs/pull/873#discussion_r1443105018

File: FIPS/fip-actor-upgrades.md (hunk 6 above)
Comment (jennijuju) [reply FIPS/fip-actor-upgrades.md:141]: discard given  https://github.com/filecoin-project/FIPs/pull/873#discussion_r1443105018

**Comment by jennijuju** [comment]
> However, I think we need more discussion on whether this is a desired feature on Filecoin network (before it can be considered for last call)& have better understanding on security implications for network application users when user actors aren't immutable. i.e: Defi and so on. See [here](https://github.com/filecoin-project/FIPs/pull/873#discussion_r1441702560)

please discard this comment - thanks to the clarification [here](https://github.com/filecoin-project/FIPs/pull/873#discussion_r1443105018)

File: FIPS/fip-actor-upgrades.md (hunk 13 above)
Comment (jsoares) [reply FIPS/fip-actor-upgrades.md:137]: @kaitlin-beegle I don't think I get why those are incentive considerations rather than product considerations.

File: FIPS/fip-actor-upgrades.md [hunk 15]
```diff
@@ -0,0 +1,153 @@
+---
+fip: "<to be assigned>" <!--keep the qoutes around the fip number, i.e: `fip: "0001"`-->
+title: Add support for upgradable actors
+author: Fridrik Asmundsson (@fridrik01), Steven Allen (@stebalien)
+discussions-to: https://github.com/filecoin-project/FIPs/discussions/396
+status: Draft
+type: Technical Core
```
Comment (jsoares) [inline FIPS/fip-actor-upgrades.md:7]:
```suggestion
type: Technical
category: Core
```

File: FIPS/fip-actor-upgrades.md (hunk 14 above)
Comment (jsoares) [reply FIPS/fip-actor-upgrades.md:13]: +1 this is not currently clear

File: FIPS/fip-actor-upgrades.md (hunk 8 above)
Comment (fridrik01) [reply FIPS/fip-actor-upgrades.md:109]: yebb

File: FIPS/fip-actor-upgrades.md (hunk 9 above)
Comment (fridrik01) [reply FIPS/fip-actor-upgrades.md:89]: This is the block returned by the upgrade endpoint, will add comments to make that explicit

**Review by jsoares — APPROVED** [review]

File: FIPS/fip-actor-upgrades.md [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+
+## Test Cases
+
+Detailed test cases are provided with the implementation.
+
+## Security Considerations
+
+Upgradable actors pose potential security risks, as users can replace deployed actors' code. However, measures are in place to minimize these risks:
+
+- Upgradable actors are opt-in by default, ensuring no impact on currently deployed actors.
```
Comment (anorth) [inline FIPS/fip-actor-upgrades.md:133]: I don't understand from the text above how this is enforced.

**Review by anorth — COMMENTED** [review]
I came away with some confusion over how this mechanism is specified to work, and thus what properties it has. Some of my comments below may make incorrect assertions, but if so these signal that the text is confusing.

**Comment by jsoares** [comment]
Assigned FIP-0088; please update the FIP frontmatter, title heading, and readme.md accordingly. @anorth's review also raises important points that should be addressed before merging.

**Comment by jsoares** [comment]
@fridrik01 Review comments have been pending for 5 weeks now. Could you clarify the status and ETA?

**Comment by fridrik01** [comment]
Closing, will open once I have addressed the comments, likely next week
